<!DOCTYPE html>
<!-- START: inst/pkgdown/templates/layout.html --><!-- Generated by pkgdown: do not edit by hand --><html lang="en" data-bs-theme="auto"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><title>Lesson Title: Section 3 : Differential Expression using a pseudobulk approach and DESeq2</title><meta name="viewport" content="width=device-width, initial-scale=1"><script src="../assets/themetoggle.js"></script><link rel="stylesheet" type="text/css" href="../assets/styles.css"><script src="../assets/scripts.js" type="text/javascript"></script><!-- mathjax --><script type="text/x-mathjax-config">
    MathJax.Hub.Config({
      config: ["MMLorHTML.js"],
      jax: ["input/TeX","input/MathML","output/HTML-CSS","output/NativeMML", "output/PreviewHTML"],
      extensions: ["tex2jax.js","mml2jax.js","MathMenu.js","MathZoom.js", "fast-preview.js", "AssistiveMML.js", "a11y/accessibility-menu.js"],
      TeX: {
        extensions: ["AMSmath.js","AMSsymbols.js","noErrors.js","noUndefined.js"]
      },
      tex2jax: {
        inlineMath: [['\\(', '\\)']],
        displayMath: [ ['$$','$$'], ['\\[', '\\]'] ],
        processEscapes: true
      }
    });
    </script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><!-- Responsive Favicon for The Carpentries --><link rel="apple-touch-icon" sizes="180x180" href="../favicons/incubator/apple-touch-icon.png"><link rel="icon" type="image/png" sizes="32x32" href="../favicons/incubator/favicon-32x32.png"><link rel="icon" type="image/png" sizes="16x16" href="../favicons/incubator/favicon-16x16.png"><link rel="manifest" href="../favicons/incubator/site.webmanifest"><link rel="mask-icon" href="../favicons/incubator/safari-pinned-tab.svg" color="#5bbad5"><meta name="msapplication-TileColor" content="#da532c"><meta name="theme-color" media="(prefers-color-scheme: light)" content="white"><meta name="theme-color" media="(prefers-color-scheme: dark)" content="black"></head><body>
    <header id="top" class="navbar navbar-expand-md top-nav incubator"><svg xmlns="http://www.w3.org/2000/svg" class="d-none"><symbol id="check2" viewbox="0 0 16 16"><path d="M13.854 3.646a.5.5 0 0 1 0 .708l-7 7a.5.5 0 0 1-.708 0l-3.5-3.5a.5.5 0 1 1 .708-.708L6.5 10.293l6.646-6.647a.5.5 0 0 1 .708 0z"></path></symbol><symbol id="circle-half" viewbox="0 0 16 16"><path d="M8 15A7 7 0 1 0 8 1v14zm0 1A8 8 0 1 1 8 0a8 8 0 0 1 0 16z"></path></symbol><symbol id="moon-stars-fill" viewbox="0 0 16 16"><path d="M6 .278a.768.768 0 0 1 .08.858 7.208 7.208 0 0 0-.878 3.46c0 4.021 3.278 7.277 7.318 7.277.527 0 1.04-.055 1.533-.16a.787.787 0 0 1 .81.316.733.733 0 0 1-.031.893A8.349 8.349 0 0 1 8.344 16C3.734 16 0 12.286 0 7.71 0 4.266 2.114 1.312 5.124.06A.752.752 0 0 1 6 .278z"></path><path d="M10.794 3.148a.217.217 0 0 1 .412 0l.387 1.162c.173.518.579.924 1.097 1.097l1.162.387a.217.217 0 0 1 0 .412l-1.162.387a1.734 1.734 0 0 0-1.097 1.097l-.387 1.162a.217.217 0 0 1-.412 0l-.387-1.162A1.734 1.734 0 0 0 9.31 6.593l-1.162-.387a.217.217 0 0 1 0-.412l1.162-.387a1.734 1.734 0 0 0 1.097-1.097l.387-1.162zM13.863.099a.145.145 0 0 1 .274 0l.258.774c.115.346.386.617.732.732l.774.258a.145.145 0 0 1 0 .274l-.774.258a1.156 1.156 0 0 0-.732.732l-.258.774a.145.145 0 0 1-.274 0l-.258-.774a1.156 1.156 0 0 0-.732-.732l-.774-.258a.145.145 0 0 1 0-.274l.774-.258c.346-.115.617-.386.732-.732L13.863.1z"></path></symbol><symbol id="sun-fill" viewbox="0 0 16 16"><path d="M8 12a4 4 0 1 0 0-8 4 4 0 0 0 0 8zM8 0a.5.5 0 0 1 .5.5v2a.5.5 0 0 1-1 0v-2A.5.5 0 0 1 8 0zm0 13a.5.5 0 0 1 .5.5v2a.5.5 0 0 1-1 0v-2A.5.5 0 0 1 8 13zm8-5a.5.5 0 0 1-.5.5h-2a.5.5 0 0 1 0-1h2a.5.5 0 0 1 .5.5zM3 8a.5.5 0 0 1-.5.5h-2a.5.5 0 0 1 0-1h2A.5.5 0 0 1 3 8zm10.657-5.657a.5.5 0 0 1 0 .707l-1.414 1.415a.5.5 0 1 1-.707-.708l1.414-1.414a.5.5 0 0 1 .707 0zm-9.193 9.193a.5.5 0 0 1 0 .707L3.05 13.657a.5.5 0 0 1-.707-.707l1.414-1.414a.5.5 0 0 1 .707 0zm9.193 2.121a.5.5 0 0 1-.707 0l-1.414-1.414a.5.5 0 0 1 .707-.707l1.414 1.414a.5.5 0 0 1 0 .707zM4.464 4.465a.5.5 0 0 1-.707 0L2.343 3.05a.5.5 0 1 1 .707-.707l1.414 1.414a.5.5 0 0 1 0 .708z"></path></symbol></svg><a class="visually-hidden-focusable skip-link" href="#main-content">Skip to main content</a>
  <div class="container-fluid top-nav-container">
    <div class="col-md-8">
      <div class="large-logo">
        <img id="incubator-logo" alt="Carpentries Incubator" src="../assets/images/incubator-logo.svg"><span class="badge text-bg-danger">
          <abbr title="This lesson is in the pre-alpha phase, which means that it is in early development, but has not yet been taught.">
            <a href="https://docs.carpentries.org/resources/curriculum/lesson-life-cycle.html" class="external-link alert-link">
              <i aria-hidden="true" class="icon" data-feather="alert-octagon" style="border-radius: 5px"></i>
              Pre-Alpha
            </a>
            <span class="visually-hidden">This lesson is in the pre-alpha phase, which means that it is in early development, but has not yet been taught.</span>
          </abbr>
        </span>

      </div>
    </div>
    <div class="selector-container">
      <div id="theme-selector">
        <li class="nav-item dropdown" id="theme-button-list">
          <button class="btn btn-link nav-link px-0 px-lg-2 dropdown-toggle d-flex align-items-center" id="bd-theme" type="button" aria-expanded="false" data-bs-toggle="dropdown" data-bs-display="static" aria-label="Toggle theme (auto)">
            <svg class="bi my-1 theme-icon-active"><use href="#circle-half"></use></svg><i data-feather="chevron-down"></i>
          </button>
          <ul class="dropdown-menu dropdown-menu-end" aria-labelledby="bd-theme-text"><li>
              <button type="button" class="btn dropdown-item d-flex align-items-center" data-bs-theme-value="light" aria-pressed="false">
                <svg class="bi me-2 theme-icon"><use href="#sun-fill"></use></svg>
                Light
                <svg class="bi ms-auto d-none"><use href="#check2"></use></svg></button>
            </li>
            <li>
              <button type="button" class="btn dropdown-item d-flex align-items-center" data-bs-theme-value="dark" aria-pressed="false">
                <svg class="bi me-2 theme-icon"><use href="#moon-stars-fill"></use></svg>
                Dark
                <svg class="bi ms-auto d-none"><use href="#check2"></use></svg></button>
            </li>
            <li>
              <button type="button" class="btn dropdown-item d-flex align-items-center active" data-bs-theme-value="auto" aria-pressed="true">
                <svg class="bi me-2 theme-icon"><use href="#circle-half"></use></svg>
                Auto
                <svg class="bi ms-auto d-none"><use href="#check2"></use></svg></button>
            </li>
          </ul></li>
      </div>

      <div class="dropdown" id="instructor-dropdown">
        <button class="btn btn-secondary dropdown-toggle bordered-button" type="button" id="dropdownMenu1" data-bs-toggle="dropdown" aria-expanded="false">
          <i aria-hidden="true" class="icon" data-feather="eye"></i> Instructor View <i data-feather="chevron-down"></i>
        </button>
        <ul class="dropdown-menu" aria-labelledby="dropdownMenu1"><li><button class="dropdown-item" type="button" onclick="window.location.href='../section3.html';">Learner View</button></li>
        </ul></div>
    </div>
  </div>
  <hr></header><nav class="navbar navbar-expand-xl bottom-nav incubator" aria-label="Main Navigation"><div class="container-fluid nav-container">
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbarSupportedContent" aria-controls="navbarSupportedContent" aria-expanded="false" aria-label="Toggle Navigation">
      <span class="navbar-toggler-icon"></span>
      <span class="menu-title">Menu</span>
    </button>
    <div class="nav-logo">
      <img class="small-logo" alt="Carpentries Incubator" src="../assets/images/incubator-logo-sm.svg"></div>
    <div class="lesson-title-md">
      Lesson Title
    </div>
    <div class="search-icon-sm">
      <!-- TODO: do not show until we have search
        <i role="img" aria-label="Search the All In One page" data-feather="search"></i>
      -->
    </div>
    <div class="desktop-nav">
      <ul class="navbar-nav me-auto mb-2 mb-lg-0"><li class="nav-item">
          <span class="lesson-title">
            Lesson Title
          </span>
        </li>
        <li class="nav-item">
          <a class="nav-link" href="../instructor/key-points.html">Key Points</a>
        </li>
        <li class="nav-item">
          <a class="nav-link" href="../instructor/instructor-notes.html">Instructor Notes</a>
        </li>
        <li class="nav-item">
          <a class="nav-link" href="../instructor/images.html">Extract All Images</a>
        </li>
        <li class="nav-item dropdown">
          <button class="nav-link dropdown-toggle" id="navbarDropdown" data-bs-toggle="dropdown" aria-expanded="false">
            More <i data-feather="chevron-down"></i>
          </button>
          <ul class="dropdown-menu" aria-labelledby="navbarDropdown"><hr><li><a class="dropdown-item" href="reference.html">Reference</a></li>
          </ul></li>
      </ul></div>
    <!--
    <form class="d-flex col-md-2 search-form">
      <fieldset disabled>
      <input class="form-control me-2 searchbox" type="search" placeholder="" aria-label="">
        <button class="btn btn-outline-success tablet-search-button"  type="submit">
          <i class="search-icon" data-feather="search" role="img" aria-label="Search the All In One page"></i>
        </button>
      </fieldset>
    </form>
    -->
    <a id="search-button" class="btn btn-primary" href="../instructor/aio.html" role="button" aria-label="Search the All In One page">Search the All In One page</a>
  </div><!--/div.container-fluid -->
</nav><div class="col-md-12 mobile-title">
  Lesson Title
</div>

<aside class="col-md-12 lesson-progress"><div style="width: 75%" class="percentage">
    75%
  </div>
  <div class="progress incubator">
    <div class="progress-bar incubator" role="progressbar" style="width: 75%" aria-valuenow="75" aria-label="Lesson Progress" aria-valuemin="0" aria-valuemax="100">
    </div>
  </div>
</aside><div class="container">
      <div class="row">
        <!-- START: inst/pkgdown/templates/navbar.html -->
<div id="sidebar-col" class="col-lg-4">
  <div id="sidebar" class="sidebar">
      <nav aria-labelledby="flush-headingEleven"><button role="button" aria-label="close menu" alt="close menu" aria-expanded="true" aria-controls="sidebar" class="collapse-toggle" data-collapse="Collapse " data-episodes="Episodes ">
          <i class="search-icon" data-feather="x" role="img"></i>
        </button>
        <div class="sidebar-inner">
          <div class="row mobile-row" id="theme-row-mobile">
            <div class="col" id="theme-selector">
              <li class="nav-item dropdown" id="theme-button-list">
                <button class="btn btn-link nav-link px-0 px-lg-2 dropdown-toggle d-flex align-items-center" id="bd-theme" type="button" aria-expanded="false" data-bs-toggle="dropdown" data-bs-display="static" aria-label="Toggle theme (auto)">
                  <svg class="bi my-1 theme-icon-active"><use href="#circle-half"></use></svg><span class="d-lg-none ms-1" id="bd-theme-text">Toggle Theme</span>
                </button>
                <ul class="dropdown-menu dropdown-menu-right" aria-labelledby="bd-theme-text"><li>
                    <button type="button" class="btn dropdown-item d-flex align-items-center" data-bs-theme-value="light" aria-pressed="false">
                      <svg class="bi me-2 theme-icon"><use href="#sun-fill"></use></svg>
                      Light
                      <svg class="bi ms-auto d-none"><use href="#check2"></use></svg></button>
                  </li>
                  <li>
                    <button type="button" class="btn dropdown-item d-flex align-items-center" data-bs-theme-value="dark" aria-pressed="false">
                      <svg class="bi me-2 theme-icon"><use href="#moon-stars-fill"></use></svg>
                      Dark
                      <svg class="bi ms-auto d-none"><use href="#check2"></use></svg></button>
                  </li>
                  <li>
                    <button type="button" class="btn dropdown-item d-flex align-items-center active" data-bs-theme-value="auto" aria-pressed="true">
                      <svg class="bi me-2 theme-icon"><use href="#circle-half"></use></svg>
                      Auto
                      <svg class="bi ms-auto d-none"><use href="#check2"></use></svg></button>
                  </li>
                </ul></li>
            </div>
          </div>
          <div class="row mobile-row">
            <div class="col">
              <div class="sidenav-view-selector">
                <div class="accordion accordion-flush" id="accordionFlush9">
                  <div class="accordion-item">
                    <h2 class="accordion-header" id="flush-headingNine">
                      <button class="accordion-button collapsed" id="instructor" type="button" data-bs-toggle="collapse" data-bs-target="#flush-collapseNine" aria-expanded="false" aria-controls="flush-collapseNine">
                        <i id="eye" aria-hidden="true" class="icon" data-feather="eye"></i> Instructor View
                      </button>
                    </h2>
                    <div id="flush-collapseNine" class="accordion-collapse collapse" aria-labelledby="flush-headingNine" data-bs-parent="#accordionFlush2">
                      <div class="accordion-body">
                        <a href="../section3.html">Learner View</a>
                      </div>
                    </div>
                  </div><!--/div.accordion-item-->
                </div><!--/div.accordion-flush-->
              </div><!--div.sidenav-view-selector -->
            </div><!--/div.col -->

            <hr></div><!--/div.mobile-row -->

          <div class="accordion accordion-flush" id="accordionFlush11">
            <div class="accordion-item">

              <button id="chapters" class="accordion-button show" type="button" data-bs-toggle="collapse" data-bs-target="#flush-collapseEleven" aria-expanded="false" aria-controls="flush-collapseEleven">
                <h2 class="accordion-header chapters" id="flush-headingEleven">
                  EPISODES
                </h2>
              </button>
              <div id="flush-collapseEleven" class="accordion-collapse show collapse" aria-labelledby="flush-headingEleven" data-bs-parent="#accordionFlush11">

                <div class="accordion-body">
                  <div class="accordion accordion-flush" id="accordionFlush1">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading1">
        <a href="index.html">Summary and Schedule</a>
    </div><!--/div.accordion-header-->

  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush2">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading2">
        <a href="introduction.html">1. Introduction</a>
    </div><!--/div.accordion-header-->

  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush3">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading3">
        <a href="section1.html">2. Section 1: Setup, Quality Control and Sample Integration</a>
    </div><!--/div.accordion-header-->

  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush4">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading4">
        <a href="section2.html">3. Section 2: Differential Gene Expression when dealing with two treatment conditions</a>
    </div><!--/div.accordion-header-->

  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlushcurrent">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-headingcurrent">
      <button class="accordion-button" type="button" data-bs-toggle="collapse" data-bs-target="#flush-collapsecurrent" aria-expanded="true" aria-controls="flush-collapsecurrent">
        <span class="visually-hidden">Current Chapter</span>
        <span class="current-chapter">
        4. Section 3 : Differential Expression using a pseudobulk approach and DESeq2
        </span>
      </button>
    </div><!--/div.accordion-header-->

    <div id="flush-collapsecurrent" class="accordion-collapse collapse show" aria-labelledby="flush-headingcurrent" data-bs-parent="#accordionFlushcurrent">
      <div class="accordion-body">
        <ul><li><a href="#section-3-differential-expression-using-a-pseudobulk-approach-and-deseq2">Section 3 : Differential Expression using a pseudobulk approach and
DESeq2</a></li>
        </ul></div><!--/div.accordion-body-->
    </div><!--/div.accordion-collapse-->

  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

                </div>
              </div>
            </div>

            <hr class="half-width"><div class="accordion accordion-flush lesson-resources" id="accordionFlush12">
              <div class="accordion-item">
                <h2 class="accordion-header" id="flush-headingTwelve">
                  <button class="accordion-button collapsed" id="lesson-resources" type="button" data-bs-toggle="collapse" data-bs-target="#flush-collapseTwelve" aria-expanded="false" aria-controls="flush-collapseTwelve">
                    RESOURCES
                  </button>
                </h2>
                <div id="flush-collapseTwelve" class="accordion-collapse collapse" aria-labelledby="flush-headingTwelve" data-bs-parent="#accordionFlush12">
                  <div class="accordion-body">
                    <ul><li>
                        <a href="../instructor/key-points.html">Key Points</a>
                      </li>
                      <li>
                        <a href="../instructor/instructor-notes.html">Instructor Notes</a>
                      </li>
                      <li>
                        <a href="../instructor/images.html">Extract All Images</a>
                      </li>
                      <hr><li><a class="dropdown-item" href="reference.html">Reference</a></li>
                    </ul></div>
                </div>
              </div>
            </div>
            <hr class="half-width lesson-resources"><a href="../instructor/aio.html">See all in one page</a>


            <hr class="d-none d-sm-block d-md-none"><div class="d-grid gap-1">

            </div>
          </div><!-- /div.accordion -->
        </div><!-- /div.sidebar-inner -->
      </nav></div><!-- /div.sidebar -->
  </div><!-- /div.sidebar-col -->
<!-- END:   inst/pkgdown/templates/navbar.html-->

        <!-- START: inst/pkgdown/templates/content-instructor.html -->
  <div class="col-xl-8 col-lg-12 primary-content">
    <nav class="lesson-content mx-md-4" aria-label="Previous and Next Chapter"><!-- content for small screens --><div class="d-block d-sm-block d-md-none">
        <a class="chapter-link" href="../instructor/section2.html"><i aria-hidden="true" class="small-arrow" data-feather="arrow-left"></i>Previous</a>
        <a class="chapter-link float-end" href="../instructor/index.html">Next<i aria-hidden="true" class="small-arrow" data-feather="arrow-right"></i></a>
      </div>
      <!-- content for large screens -->
      <div class="d-none d-sm-none d-md-block">
        <a class="chapter-link" href="../instructor/section2.html" rel="prev">
          <i aria-hidden="true" class="small-arrow" data-feather="arrow-left"></i>
          Previous: Section 2:
        </a>
        <a class="chapter-link float-end" href="../instructor/index.html" rel="next">
          Home
          <i aria-hidden="true" class="small-arrow" data-feather="arrow-right"></i>
        </a>
      </div>
      <hr></nav><main id="main-content" class="main-content"><div class="container lesson-content">
        <h1>Section 3 : Differential Expression using a pseudobulk approach and DESeq2</h1>
        <p>Last updated on 2025-11-02 |

        <a href="https://example.com/template-source/lesson-example/edit/main/episodes/section3.Rmd" class="external-link">Edit this page <i aria-hidden="true" data-feather="edit"></i></a></p>



        <p>Estimated time: <i aria-hidden="true" data-feather="clock"></i> 12 minutes</p>

        <div class="text-end">
          <button role="button" aria-pressed="false" tabindex="0" id="expand-code" class="pull-right" data-expand="Expand All Solutions " data-collapse="Collapse All Solutions "> Expand All Solutions <i aria-hidden="true" data-feather="plus"></i></button>
        </div>

        

<div class="overview card">
<h2 class="card-header">Overview</h2>
<div class="row g-0">
<div class="col-md-4">
<div class="card-body">
<div class="inner">
<h3 class="card-title">Questions</h3>
<ul><li>Why aggregate to pseudobulk instead of testing each cell?</li>
<li>How does donor act as the biological replicate in DESeq2?</li>
<li>When do pseudobulk and single-cell DE agree or disagree, and why
(e.g., zero inflation, composition, variance models)?</li>
<li>What minimum sampling per group (cells per donor, donors per group)
is sensible before aggregating?</li>
</ul></div>
</div>
</div>
<div class="col-md-8">
<div class="card-body">
<div class="inner bordered">
<h3 class="card-title">Objectives</h3>
<ul><li>Attach donor IDs to cells and create pseudobulk matrices with
AggregateExpression().</li>
<li>Run DESeq2 from Seurat via FindMarkers(test.use=“DESeq2”) on
pseudobulk objects.</li>
<li>Inspect and interpret key outputs (adjusted p-values, log2FC) and
compare to single-cell DE.</li>
<li>Visualize consensus/unique DEGs and make a publication-ready heatmap
with pheatmap.</li>
<li>Recognize pitfalls (uneven donor coverage, tiny groups, composition
shifts) and report limitations.</li>
</ul></div>
</div>
</div>
</div>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>Modularity Optimizer version 1.3.0 by Ludo Waltman and Nees Jan van Eck

Number of nodes: 13548
Number of edges: 521570

Running Louvain algorithm...
Maximum modularity in 10 random starts: 0.9002
Number of communities: 13
Elapsed time: 1 seconds</code></pre>
</div>
<section><h2 class="section-heading" id="section-3-differential-expression-using-a-pseudobulk-approach-and-deseq2">Section 3 : Differential Expression using a pseudobulk approach and
DESeq2<a class="anchor" aria-label="anchor" href="#section-3-differential-expression-using-a-pseudobulk-approach-and-deseq2"></a></h2>
<hr class="half-width"><div class="section level3">
<h3 id="step-1-we-need-to-import-sample-information-for-each-cell-from-the-original-paper">Step 1: We need to import sample information for each cell from the
original paper<a class="anchor" aria-label="anchor" href="#step-1-we-need-to-import-sample-information-for-each-cell-from-the-original-paper"></a></h3>
<div id="discussion1" class="callout discussion">
<div class="callout-square">
<i class="callout-icon" data-feather="message-circle"></i>
</div>
<span class="callout-header">Discussion</span>
<div class="callout-inner">
<div class="callout-content">
<p>Have a look at the ifnb.filtered seurat metadata, can you spot what
we’ve done here?</p>
</div>
</div>
</div>
<div class="codewrapper sourceCode" id="cb2">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># defining a function here to retrieve that information (code from https://satijalab.org/seurat/articles/de_vignette)</span></span>
<span><span class="va">loadDonorMetadata</span> <span class="op">&lt;-</span> <span class="kw">function</span><span class="op">(</span><span class="va">seu.obj</span><span class="op">)</span><span class="op">{</span></span>
<span>  <span class="co"># load the inferred sample IDs of each cell</span></span>
<span>  <span class="va">ctrl</span> <span class="op">&lt;-</span> <span class="fu">read.table</span><span class="op">(</span><span class="fu">url</span><span class="op">(</span><span class="st">"https://raw.githubusercontent.com/yelabucsf/demuxlet_paper_code/master/fig3/ye1.ctrl.8.10.sm.best"</span><span class="op">)</span>, head <span class="op">=</span> <span class="cn">T</span>, stringsAsFactors <span class="op">=</span> <span class="cn">F</span><span class="op">)</span></span>
<span>  <span class="va">stim</span> <span class="op">&lt;-</span> <span class="fu">read.table</span><span class="op">(</span><span class="fu">url</span><span class="op">(</span><span class="st">"https://raw.githubusercontent.com/yelabucsf/demuxlet_paper_code/master/fig3/ye2.stim.8.10.sm.best"</span><span class="op">)</span>, head <span class="op">=</span> <span class="cn">T</span>, stringsAsFactors <span class="op">=</span> <span class="cn">F</span><span class="op">)</span></span>
<span>  <span class="va">info</span> <span class="op">&lt;-</span> <span class="fu">rbind</span><span class="op">(</span><span class="va">ctrl</span>, <span class="va">stim</span><span class="op">)</span></span>
<span>  </span>
<span>  <span class="co"># rename the cell IDs by substituting the '-' into '.'</span></span>
<span>  <span class="va">info</span><span class="op">$</span><span class="va">BARCODE</span> <span class="op">&lt;-</span> <span class="fu">gsub</span><span class="op">(</span>pattern <span class="op">=</span> <span class="st">"\\-"</span>, replacement <span class="op">=</span> <span class="st">"\\."</span>, <span class="va">info</span><span class="op">$</span><span class="va">BARCODE</span><span class="op">)</span></span>
<span>  </span>
<span>  <span class="co"># only keep the cells with high-confidence sample ID</span></span>
<span>  <span class="va">info</span> <span class="op">&lt;-</span> <span class="va">info</span><span class="op">[</span><span class="fu">grep</span><span class="op">(</span>pattern <span class="op">=</span> <span class="st">"SNG"</span>, x <span class="op">=</span> <span class="va">info</span><span class="op">$</span><span class="va">BEST</span><span class="op">)</span>, <span class="op">]</span></span>
<span>  </span>
<span>  <span class="co"># remove cells with duplicated IDs in both ctrl and stim groups</span></span>
<span>  <span class="va">info</span> <span class="op">&lt;-</span> <span class="va">info</span><span class="op">[</span><span class="op">!</span><span class="fu">duplicated</span><span class="op">(</span><span class="va">info</span><span class="op">$</span><span class="va">BARCODE</span><span class="op">)</span> <span class="op">&amp;</span> <span class="op">!</span><span class="fu">duplicated</span><span class="op">(</span><span class="va">info</span><span class="op">$</span><span class="va">BARCODE</span>, fromLast <span class="op">=</span> <span class="cn">T</span><span class="op">)</span>, <span class="op">]</span></span>
<span>  </span>
<span>  <span class="co"># now add the sample IDs to ifnb </span></span>
<span>  <span class="fu">rownames</span><span class="op">(</span><span class="va">info</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="va">info</span><span class="op">$</span><span class="va">BARCODE</span></span>
<span>  <span class="va">info</span> <span class="op">&lt;-</span> <span class="va">info</span><span class="op">[</span>, <span class="fu">c</span><span class="op">(</span><span class="st">"BEST"</span><span class="op">)</span>, drop <span class="op">=</span> <span class="cn">F</span><span class="op">]</span></span>
<span>  <span class="fu">names</span><span class="op">(</span><span class="va">info</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu">c</span><span class="op">(</span><span class="st">"donor_id"</span><span class="op">)</span></span>
<span>  <span class="va">seu.obj</span> <span class="op">&lt;-</span> <span class="fu">AddMetaData</span><span class="op">(</span><span class="va">seu.obj</span>, metadata <span class="op">=</span> <span class="va">info</span><span class="op">)</span></span>
<span>  </span>
<span>  <span class="co"># remove cells without donor IDs</span></span>
<span>  <span class="va">seu.obj</span><span class="op">$</span><span class="va">donor_id</span><span class="op">[</span><span class="fu">is.na</span><span class="op">(</span><span class="va">seu.obj</span><span class="op">$</span><span class="va">donor_id</span><span class="op">)</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="st">"unknown"</span></span>
<span>  <span class="va">seu.obj</span> <span class="op">&lt;-</span> <span class="fu">subset</span><span class="op">(</span><span class="va">seu.obj</span>, subset <span class="op">=</span> <span class="va">donor_id</span> <span class="op">!=</span> <span class="st">"unknown"</span><span class="op">)</span></span>
<span>  </span>
<span>  <span class="kw">return</span><span class="op">(</span><span class="va">seu.obj</span><span class="op">)</span></span>
<span><span class="op">}</span></span>
<span></span>
<span><span class="va">ifnb.filtered</span> <span class="op">&lt;-</span> <span class="fu">loadDonorMetadata</span><span class="op">(</span><span class="va">ifnb.filtered</span><span class="op">)</span></span></code></pre>
</div>
</div>
<div class="section level3">
<h3 id="step-2-aggregate-our-counts-based-on-treatment-group-cell-type-and-donor-id">Step 2: Aggregate our counts based on treatment group, cell-type,
and donor id<a class="anchor" aria-label="anchor" href="#step-2-aggregate-our-counts-based-on-treatment-group-cell-type-and-donor-id"></a></h3>
<p>Collapsing our single-cell matrix in this manner is referred to as
creating a ‘pseudo-bulk’ count matrix.. We’re making a condensed count
matrix that looks more like a bulk matrix so that we can use bulk
differential expression algorithms like DESeq2. We can see this clearly
when we have a look at the ifnb.pseudbulk.df we make in the following
code block.</p>
<div class="codewrapper sourceCode" id="cb3">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">ifnb.pseudobulk</span> <span class="op">&lt;-</span> <span class="fu">AggregateExpression</span><span class="op">(</span><span class="va">ifnb.filtered</span>, assays <span class="op">=</span> <span class="st">"RNA"</span>,</span>
<span>                                   group.by <span class="op">=</span> <span class="fu">c</span><span class="op">(</span><span class="st">"stim"</span>, <span class="st">"donor_id"</span>, <span class="st">"seurat_annotations"</span><span class="op">)</span>,</span>
<span>                                   return.seurat <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>Centering and scaling data matrix</code></pre>
</div>
<div class="codewrapper sourceCode" id="cb5">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## Centering and scaling data matrix</span></span>
<span><span class="co"># If you want the pseudobulk matrix as a dataframe you can do this:</span></span>
<span><span class="va">ifnb.pseudobulk.df</span> <span class="op">&lt;-</span> <span class="fu">AggregateExpression</span><span class="op">(</span><span class="va">ifnb.filtered</span>, assays <span class="op">=</span> <span class="st">"RNA"</span>,</span>
<span>                                          group.by <span class="op">=</span> <span class="fu">c</span><span class="op">(</span><span class="st">"stim"</span>, <span class="st">"donor_id"</span>, <span class="st">"seurat_annotations"</span><span class="op">)</span><span class="op">)</span> <span class="op">%&gt;%</span> </span>
<span>  <span class="fu">as.data.frame</span><span class="op">(</span><span class="op">)</span></span></code></pre>
</div>
<p>We can view the top rows here:</p>
<div class="codewrapper sourceCode" id="cb6">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">head</span><span class="op">(</span><span class="va">ifnb.pseudobulk.df</span><span class="op">)</span></span></code></pre>
</div>
</div>
<div class="section level3">
<h3 id="step-3-perform-differential-expression-using-deseq2">Step 3: Perform Differential Expression using DESeq2<a class="anchor" aria-label="anchor" href="#step-3-perform-differential-expression-using-deseq2"></a></h3>
<p>Just like before, lets make a new column containing the cell type and
treatment group before DE with DESeq2. We do this so that we can
leverage the Idents() function combined with the FindMarkers
function.</p>
<p>To use DESeq2 in older versions of Seurat (prior to v5) we would have
to perform a myriad of convoluted data wrangling steps to convert our
Seurat pseudobulk matrix into a format compatible with DESeq2 (or
whatever bulk DE package being used). Fortunately in Seurat v5, these
different bulk DE algorithms/packages have been wrapped into the
FindMarker() functions - simplifying pseudobulk DE approaches
immensely.</p>
<p>It is worth taking a look at the Seurat DE vignette on your own to
see the other pseudobulk DE methods you can use. Do keep in mind that to
use some of the DE tests, you need to install that package and load it
as a library separately to Seurat, as we’ve done here with DESeq2 at the
start.</p>
<div class="codewrapper sourceCode" id="cb7">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">ifnb.pseudobulk</span><span class="op">$</span><span class="va">celltype.and.stim</span> <span class="op">&lt;-</span> <span class="fu">paste</span><span class="op">(</span><span class="va">ifnb.pseudobulk</span><span class="op">$</span><span class="va">seurat_annotations</span>, <span class="va">ifnb.pseudobulk</span><span class="op">$</span><span class="va">stim</span>, sep <span class="op">=</span> <span class="st">"_"</span><span class="op">)</span></span>
<span><span class="fu">Idents</span><span class="op">(</span><span class="va">ifnb.pseudobulk</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="st">"celltype.and.stim"</span></span>
<span></span>
<span></span>
<span><span class="co"># Lets run a DEG test between treated and control CD 16 monocytes using the same FindMarkers function but with DESeq2</span></span>
<span><span class="va">treatment.response.CD16.pseudo</span> <span class="op">&lt;-</span> <span class="fu">FindMarkers</span><span class="op">(</span>object <span class="op">=</span> <span class="va">ifnb.pseudobulk</span>, </span>
<span>                                      ident.1 <span class="op">=</span> <span class="st">'CD16 Mono_STIM'</span>, </span>
<span>                                      ident.2 <span class="op">=</span> <span class="st">'CD16 Mono_CTRL'</span>,</span>
<span>                                      test.use <span class="op">=</span> <span class="st">"DESeq2"</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>converting counts to integer mode</code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>gene-wise dispersion estimates</code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>mean-dispersion relationship</code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>final dispersion estimates</code></pre>
</div>
<div class="codewrapper sourceCode" id="cb12">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">head</span><span class="op">(</span><span class="va">treatment.response.CD16.pseudo</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>               p_val avg_log2FC pct.1 pct.2     p_val_adj
IFIT3  1.564083e-134   4.601427     1     1 2.198006e-130
IFIT2   2.122691e-84   4.613021     1     1  2.983017e-80
ISG20   1.401656e-81   4.038272     1     1  1.969747e-77
DDX58   1.366535e-73   3.448721     1     1  1.920392e-69
NT5C3A  5.127048e-66   3.942571     1     1  7.205040e-62
OASL    1.186412e-63   4.025025     1     1  1.667265e-59</code></pre>
</div>
<div class="codewrapper sourceCode" id="cb14">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># How are we able to use the same findmarkers function here?</span></span>
<span><span class="fu">head</span><span class="op">(</span><span class="fu">Cells</span><span class="op">(</span><span class="va">ifnb.pseudobulk</span><span class="op">)</span><span class="op">)</span> <span class="co"># our 'cells' are no longer barcodes, but have been renamed according to stim-donor-annotation when we aggregated our data earlier</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>[1] "CTRL_SNG-101_CD14 Mono"    "CTRL_SNG-101_CD4 Naive T"
[3] "CTRL_SNG-101_CD4 Memory T" "CTRL_SNG-101_CD16 Mono"
[5] "CTRL_SNG-101_B"            "CTRL_SNG-101_CD8 T"       </code></pre>
</div>
</div>
<div class="section level3">
<h3 id="step-4-assessing-differences-between-our-pseudbulk-degs-and-single-cell-degs">Step 4: Assessing differences between our pseudbulk DEGs and
single-cell DEGs<a class="anchor" aria-label="anchor" href="#step-4-assessing-differences-between-our-pseudbulk-degs-and-single-cell-degs"></a></h3>
<p>First lets take a look at the DEG dataframes we made for both.</p>
<div id="discussion2" class="callout discussion">
<div class="callout-square">
<i class="callout-icon" data-feather="message-circle"></i>
</div>
<span class="callout-header">Discussion</span>
<div class="callout-inner">
<div class="callout-content">
<p>Having a look at them, can you identify any differences? Can you
think of any underlying reasons behind those differences?</p>
<p>Hint: Look at the p_val and p_val_adj columns.</p>
</div>
</div>
</div>
<div class="codewrapper sourceCode" id="cb16">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">head</span><span class="op">(</span><span class="va">treatment.response.CD16</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>               p_val avg_log2FC pct.1 pct.2     p_val_adj
IFIT1  1.379187e-176   5.834216 1.000 0.094 1.938172e-172
ISG15  6.273887e-166   5.333771 1.000 0.478 8.816694e-162
IFIT3  1.413978e-164   4.412990 0.992 0.314 1.987063e-160
ISG20  6.983755e-164   4.088510 1.000 0.448 9.814270e-160
IFITM3 1.056793e-161   3.191513 1.000 0.634 1.485111e-157
IFIT2  7.334976e-159   4.622453 0.974 0.162 1.030784e-154</code></pre>
</div>
<div class="codewrapper sourceCode" id="cb18">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">head</span><span class="op">(</span><span class="va">treatment.response.CD16.pseudo</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>               p_val avg_log2FC pct.1 pct.2     p_val_adj
IFIT3  1.564083e-134   4.601427     1     1 2.198006e-130
IFIT2   2.122691e-84   4.613021     1     1  2.983017e-80
ISG20   1.401656e-81   4.038272     1     1  1.969747e-77
DDX58   1.366535e-73   3.448721     1     1  1.920392e-69
NT5C3A  5.127048e-66   3.942571     1     1  7.205040e-62
OASL    1.186412e-63   4.025025     1     1  1.667265e-59</code></pre>
</div>
<p>Next let’s take a look at the degree of overlap between the actual
DEGs in both approaches.</p>
<p>To help with this, I’m just going to define a couple helper functions
below.</p>
<div class="codewrapper sourceCode" id="cb20">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">Merge_DEG_dataframes</span> <span class="op">&lt;-</span> <span class="kw">function</span><span class="op">(</span><span class="va">pseudobulk.de</span>,</span>
<span>                                 <span class="va">singlecell.de</span><span class="op">)</span><span class="op">{</span></span>
<span>  <span class="fu">names</span><span class="op">(</span><span class="va">pseudobulk.de</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu">paste0</span><span class="op">(</span><span class="fu">names</span><span class="op">(</span><span class="va">pseudobulk.de</span><span class="op">)</span>, <span class="st">".bulk"</span><span class="op">)</span></span>
<span>  <span class="va">pseudobulk.de</span><span class="op">$</span><span class="va">gene</span> <span class="op">&lt;-</span> <span class="fu">rownames</span><span class="op">(</span><span class="va">pseudobulk.de</span><span class="op">)</span></span>
<span>  </span>
<span>  <span class="fu">names</span><span class="op">(</span><span class="va">singlecell.de</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu">paste0</span><span class="op">(</span><span class="fu">names</span><span class="op">(</span><span class="va">singlecell.de</span><span class="op">)</span>, <span class="st">".sc"</span><span class="op">)</span></span>
<span>  <span class="va">singlecell.de</span><span class="op">$</span><span class="va">gene</span> <span class="op">&lt;-</span> <span class="fu">rownames</span><span class="op">(</span><span class="va">singlecell.de</span><span class="op">)</span></span>
<span>  </span>
<span>  <span class="va">merge_dat</span> <span class="op">&lt;-</span> <span class="fu">merge</span><span class="op">(</span><span class="va">singlecell.de</span>, <span class="va">pseudobulk.de</span>, by <span class="op">=</span> <span class="st">"gene"</span><span class="op">)</span></span>
<span>  <span class="va">merge_dat</span> <span class="op">&lt;-</span> <span class="va">merge_dat</span><span class="op">[</span><span class="fu">order</span><span class="op">(</span><span class="va">merge_dat</span><span class="op">$</span><span class="va">p_val.bulk</span><span class="op">)</span>, <span class="op">]</span></span>
<span>  </span>
<span>  <span class="kw">return</span><span class="op">(</span><span class="va">merge_dat</span><span class="op">)</span></span>
<span><span class="op">}</span></span>
<span><span class="va">Visualise_Overlapping_DEGs</span> <span class="op">&lt;-</span> <span class="kw">function</span><span class="op">(</span><span class="va">pseudobulk.de</span>,</span>
<span>                                       <span class="va">singlecell.de</span><span class="op">)</span><span class="op">{</span></span>
<span>  </span>
<span>  <span class="va">merge_dat</span> <span class="op">&lt;-</span> <span class="fu">Merge_DEG_dataframes</span><span class="op">(</span><span class="va">pseudobulk.de</span>,</span>
<span>                                    <span class="va">singlecell.de</span><span class="op">)</span></span>
<span>  </span>
<span>  <span class="co"># Number of genes that are marginally significant in both; marginally significant only in bulk; and marginally     significant only in single-cell</span></span>
<span>  <span class="va">common</span> <span class="op">&lt;-</span> <span class="va">merge_dat</span><span class="op">$</span><span class="va">gene</span><span class="op">[</span><span class="fu">which</span><span class="op">(</span><span class="va">merge_dat</span><span class="op">$</span><span class="va">p_val_adj.bulk</span> <span class="op">&lt;</span> <span class="fl">0.05</span> <span class="op">&amp;</span> </span>
<span>                                   <span class="va">merge_dat</span><span class="op">$</span><span class="va">p_val_adj.sc</span> <span class="op">&lt;</span> <span class="fl">0.05</span><span class="op">)</span><span class="op">]</span></span>
<span>  <span class="va">only_sc</span> <span class="op">&lt;-</span> <span class="va">merge_dat</span><span class="op">$</span><span class="va">gene</span><span class="op">[</span><span class="fu">which</span><span class="op">(</span><span class="va">merge_dat</span><span class="op">$</span><span class="va">p_val_adj.bulk</span> <span class="op">&gt;</span> <span class="fl">0.05</span> <span class="op">&amp;</span> </span>
<span>                                    <span class="va">merge_dat</span><span class="op">$</span><span class="va">p_val_adj.sc</span> <span class="op">&lt;</span> <span class="fl">0.05</span><span class="op">)</span><span class="op">]</span></span>
<span>  <span class="va">only_pseudobulk</span> <span class="op">&lt;-</span> <span class="va">merge_dat</span><span class="op">$</span><span class="va">gene</span><span class="op">[</span><span class="fu">which</span><span class="op">(</span><span class="va">merge_dat</span><span class="op">$</span><span class="va">p_val_adj.bulk</span> <span class="op">&lt;</span> <span class="fl">0.05</span> <span class="op">&amp;</span> </span>
<span>                                      <span class="va">merge_dat</span><span class="op">$</span><span class="va">p_val_adj.sc</span> <span class="op">&gt;</span> <span class="fl">0.05</span><span class="op">)</span><span class="op">]</span></span>
<span>  </span>
<span>  <span class="co"># Create a dataframe to plot overlaps</span></span>
<span>  <span class="va">overlap.info</span> <span class="op">&lt;-</span> <span class="fu">data.frame</span><span class="op">(</span></span>
<span>    category <span class="op">=</span> <span class="fu">c</span><span class="op">(</span><span class="st">"Common"</span>, </span>
<span>                 <span class="st">"Only in single-cell"</span>, </span>
<span>                 <span class="st">"Only in pseudobulk"</span><span class="op">)</span>,</span>
<span>    count <span class="op">=</span> <span class="fu">c</span><span class="op">(</span><span class="fu">length</span><span class="op">(</span><span class="va">common</span><span class="op">)</span>, <span class="fu">length</span><span class="op">(</span><span class="va">only_sc</span><span class="op">)</span>, <span class="fu">length</span><span class="op">(</span><span class="va">only_pseudobulk</span><span class="op">)</span><span class="op">)</span></span>
<span>  <span class="op">)</span></span>
<span>  </span>
<span>  <span class="va">overlap.info</span><span class="op">$</span><span class="va">category</span> <span class="op">&lt;-</span> <span class="fu">factor</span><span class="op">(</span><span class="va">overlap.info</span><span class="op">$</span><span class="va">category</span>, </span>
<span>                                  levels <span class="op">=</span> <span class="fu">c</span><span class="op">(</span><span class="st">"Common"</span>, </span>
<span>                                             <span class="st">"Only in single-cell"</span>, </span>
<span>                                             <span class="st">"Only in pseudobulk"</span><span class="op">)</span><span class="op">)</span></span>
<span>  <span class="co"># Create the bar plot</span></span>
<span>  <span class="va">overlap.bar.plt</span> <span class="op">&lt;-</span> <span class="fu">ggplot</span><span class="op">(</span><span class="va">overlap.info</span>, <span class="fu">aes</span><span class="op">(</span>x <span class="op">=</span> <span class="va">category</span>, y <span class="op">=</span> <span class="va">count</span>, fill <span class="op">=</span> <span class="va">category</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu">geom_bar</span><span class="op">(</span>stat <span class="op">=</span> <span class="st">"identity"</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu">geom_text</span><span class="op">(</span><span class="fu">aes</span><span class="op">(</span>label <span class="op">=</span> <span class="va">count</span><span class="op">)</span>, vjust <span class="op">=</span> <span class="op">-</span><span class="fl">0.5</span>, size <span class="op">=</span> <span class="fl">4</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu">theme_minimal</span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu">labs</span><span class="op">(</span>title <span class="op">=</span> <span class="st">"Number of Overlapping and Unique DEGs from Single-Cell and Pseudobulk Tests"</span>,</span>
<span>         x <span class="op">=</span> <span class="st">""</span>,</span>
<span>         y <span class="op">=</span> <span class="st">"Number of Genes"</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu">theme</span><span class="op">(</span>legend.position <span class="op">=</span> <span class="st">"none"</span>,</span>
<span>          axis.text.x <span class="op">=</span> <span class="fu">element_text</span><span class="op">(</span>angle <span class="op">=</span> <span class="fl">45</span>, hjust <span class="op">=</span> <span class="fl">1</span><span class="op">)</span><span class="op">)</span></span>
<span>  </span>
<span>  <span class="kw">return</span><span class="op">(</span><span class="va">overlap.bar.plt</span><span class="op">)</span></span>
<span><span class="op">}</span></span></code></pre>
</div>
<p>Let’s use the helper functions above to plot and view the
overlap/agreement between our pseudobulk versus single-cell
approaches</p>
<div class="codewrapper sourceCode" id="cb21">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">merged_deg_data</span> <span class="op">&lt;-</span> <span class="fu">Merge_DEG_dataframes</span><span class="op">(</span>pseudobulk.de <span class="op">=</span> <span class="va">treatment.response.CD16.pseudo</span>,</span>
<span>                                        singlecell.de <span class="op">=</span> <span class="va">treatment.response.CD16</span><span class="op">)</span></span>
<span><span class="va">merged_deg_data</span> <span class="op">%&gt;%</span> </span>
<span>  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu">select</span><span class="op">(</span><span class="va">gene</span>, </span>
<span>                <span class="va">p_val.sc</span>, <span class="va">p_val.bulk</span>,</span>
<span>                <span class="va">p_val_adj.sc</span>, <span class="va">p_val_adj.bulk</span><span class="op">)</span> <span class="op">%&gt;%</span></span>
<span>  <span class="fu">head</span><span class="op">(</span><span class="fl">10</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>       gene      p_val.sc    p_val.bulk  p_val_adj.sc p_val_adj.bulk
2866  IFIT3 1.413978e-164 1.564083e-134 1.987063e-160  2.198006e-130
2865  IFIT2 7.334976e-159  2.122691e-84 1.030784e-154   2.983017e-80
2992  ISG20 6.983755e-164  1.401656e-81 9.814270e-160   1.969747e-77
1626  DDX58 2.340153e-109  1.366535e-73 3.288617e-105   1.920392e-69
4129 NT5C3A 5.806396e-117  5.127048e-66 8.159728e-113   7.205040e-62
4185   OASL 5.497910e-147  1.186412e-63 7.726213e-143   1.667265e-59
4544 PLSCR1 6.905174e-116  3.783863e-61 9.703841e-112   5.317463e-57
3859 MYL12A  2.476988e-83  7.623767e-61  3.480912e-79   1.071368e-56
2859  IFI35  4.701828e-99  2.527378e-58  6.607479e-95   3.551724e-54
2661  HERC5  1.578848e-92  1.829060e-56  2.218755e-88   2.570378e-52</code></pre>
</div>
<div class="codewrapper sourceCode" id="cb23">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># How many DEGs overlap between our two methods? Is there anything in the merged_deg_data frame that stands out to you?</span></span>
<span><span class="va">overlap.bar.plt</span> <span class="op">&lt;-</span> <span class="fu">Visualise_Overlapping_DEGs</span><span class="op">(</span>pseudobulk.de <span class="op">=</span> <span class="va">treatment.response.CD16.pseudo</span>,</span>
<span>                                              singlecell.de <span class="op">=</span> <span class="va">treatment.response.CD16</span><span class="op">)</span></span>
<span></span>
<span><span class="va">overlap.bar.plt</span></span></code></pre>
</div>
<figure><img src="../fig/section3-rendered-unnamed-chunk-8-1.png" style="display: block; margin: auto;" class="figure mx-auto d-block"></figure><div id="discussion3" class="callout discussion">
<div class="callout-square">
<i class="callout-icon" data-feather="message-circle"></i>
</div>
<span class="callout-header">Discussion</span>
<div class="callout-inner">
<div class="callout-content">
<p>Can you explain why we’re seeing the discrepancies we see between the
two methods here?</p>
</div>
</div>
</div>
</div>
<div class="section level3">
<h3 id="step-5-investigate-the-differences-between-pseudobulk-de-and-single-cell-de-closer">Step 5: Investigate the differences between pseudobulk DE and
single-cell DE closer<a class="anchor" aria-label="anchor" href="#step-5-investigate-the-differences-between-pseudobulk-de-and-single-cell-de-closer"></a></h3>
<p>Let’s create lists of genes in each of our categories first</p>
<div class="codewrapper sourceCode" id="cb24">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">common</span> <span class="op">&lt;-</span> <span class="va">merged_deg_data</span><span class="op">$</span><span class="va">gene</span><span class="op">[</span><span class="fu">which</span><span class="op">(</span><span class="va">merged_deg_data</span><span class="op">$</span><span class="va">p_val.bulk</span> <span class="op">&lt;</span> <span class="fl">0.05</span> <span class="op">&amp;</span> </span>
<span>                                       <span class="va">merged_deg_data</span><span class="op">$</span><span class="va">p_val.sc</span> <span class="op">&lt;</span> <span class="fl">0.05</span><span class="op">)</span><span class="op">]</span></span>
<span><span class="va">only_sc</span> <span class="op">&lt;-</span> <span class="va">merged_deg_data</span><span class="op">$</span><span class="va">gene</span><span class="op">[</span><span class="fu">which</span><span class="op">(</span><span class="va">merged_deg_data</span><span class="op">$</span><span class="va">p_val.bulk</span> <span class="op">&gt;</span> <span class="fl">0.05</span> <span class="op">&amp;</span> </span>
<span>                                        <span class="va">merged_deg_data</span><span class="op">$</span><span class="va">p_val.sc</span> <span class="op">&lt;</span> <span class="fl">0.05</span><span class="op">)</span><span class="op">]</span></span>
<span><span class="va">only_pseudobulk</span> <span class="op">&lt;-</span> <span class="va">merged_deg_data</span><span class="op">$</span><span class="va">gene</span><span class="op">[</span><span class="fu">which</span><span class="op">(</span><span class="va">merged_deg_data</span><span class="op">$</span><span class="va">p_val.bulk</span> <span class="op">&lt;</span> <span class="fl">0.05</span> <span class="op">&amp;</span> </span>
<span>                                          <span class="va">merged_deg_data</span><span class="op">$</span><span class="va">p_val.sc</span> <span class="op">&gt;</span> <span class="fl">0.05</span><span class="op">)</span><span class="op">]</span></span></code></pre>
</div>
<p>Now I want to look at the expression of genes that only appear in our
sc deg test and not in our pseudobulk deg test I’ve picked two genes
from the ‘only_sc’ variable we just defined</p>
<div class="codewrapper sourceCode" id="cb25">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># create a new column to annotate sample-condition-celltype in the single-cell dataset</span></span>
<span><span class="va">ifnb.filtered</span><span class="op">$</span><span class="va">donor_id.and.stim</span> <span class="op">&lt;-</span> <span class="fu">paste0</span><span class="op">(</span><span class="va">ifnb.filtered</span><span class="op">$</span><span class="va">stim</span>, <span class="st">"-"</span>, <span class="va">ifnb.filtered</span><span class="op">$</span><span class="va">donor_id</span><span class="op">)</span></span>
<span><span class="fu">Idents</span><span class="op">(</span><span class="va">ifnb.filtered</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="st">"celltype.and.stim"</span></span>
<span></span>
<span><span class="co"># Explore some genes that only appear in the sc deg test---</span></span>
<span><span class="fu">VlnPlot</span><span class="op">(</span><span class="va">ifnb.filtered</span>, features <span class="op">=</span> <span class="fu">c</span><span class="op">(</span><span class="st">"PABPC1"</span>, <span class="st">"SRGN"</span><span class="op">)</span>, </span>
<span>        idents <span class="op">=</span> <span class="fu">c</span><span class="op">(</span><span class="st">"CD16 Mono_CTRL"</span>, <span class="st">"CD16 Mono_STIM"</span><span class="op">)</span>, </span>
<span>        group.by <span class="op">=</span> <span class="st">"stim"</span><span class="op">)</span> </span></code></pre>
</div>
<figure><img src="../fig/section3-rendered-unnamed-chunk-10-1.png" style="display: block; margin: auto;" class="figure mx-auto d-block"></figure><div class="codewrapper sourceCode" id="cb26">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">VlnPlot</span><span class="op">(</span><span class="va">ifnb.filtered</span>, features <span class="op">=</span> <span class="fu">c</span><span class="op">(</span><span class="st">"PABPC1"</span>, <span class="st">"SRGN"</span><span class="op">)</span>, </span>
<span>        idents <span class="op">=</span> <span class="fu">c</span><span class="op">(</span><span class="st">"CD16 Mono_CTRL"</span>, <span class="st">"CD16 Mono_STIM"</span><span class="op">)</span>, </span>
<span>        group.by <span class="op">=</span> <span class="st">"donor_id.and.stim"</span>, ncol <span class="op">=</span> <span class="fl">1</span><span class="op">)</span></span></code></pre>
</div>
<figure><img src="../fig/section3-rendered-unnamed-chunk-11-1.png" style="display: block; margin: auto;" class="figure mx-auto d-block"></figure><div id="discussion4" class="callout discussion">
<div class="callout-square">
<i class="callout-icon" data-feather="message-circle"></i>
</div>
<span class="callout-header">Discussion</span>
<div class="callout-inner">
<div class="callout-content">
<p>How would you interpret the plots above? What does this tell you
about some of the pitfalls of single-cell DE approaches?</p>
</div>
</div>
</div>
<p>Now lets take a look at some genes that show agreement across both sc
and pseudobulk deg tests</p>
<div class="codewrapper sourceCode" id="cb27">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">VlnPlot</span><span class="op">(</span><span class="va">ifnb.filtered</span>, features <span class="op">=</span> <span class="fu">c</span><span class="op">(</span><span class="st">"IFIT2"</span>, <span class="st">"PSMA4"</span><span class="op">)</span>, </span>
<span>        idents <span class="op">=</span> <span class="fu">c</span><span class="op">(</span><span class="st">"CD16 Mono_CTRL"</span>, <span class="st">"CD16 Mono_STIM"</span><span class="op">)</span>, </span>
<span>        group.by <span class="op">=</span> <span class="st">"stim"</span><span class="op">)</span> </span></code></pre>
</div>
<figure><img src="../fig/section3-rendered-unnamed-chunk-12-1.png" style="display: block; margin: auto;" class="figure mx-auto d-block"></figure><div class="codewrapper sourceCode" id="cb28">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">VlnPlot</span><span class="op">(</span><span class="va">ifnb.filtered</span>, features <span class="op">=</span> <span class="fu">c</span><span class="op">(</span><span class="st">"IFIT2"</span>, <span class="st">"PSMA4"</span><span class="op">)</span>, </span>
<span>        idents <span class="op">=</span> <span class="fu">c</span><span class="op">(</span><span class="st">"CD16 Mono_CTRL"</span>, <span class="st">"CD16 Mono_STIM"</span><span class="op">)</span>, </span>
<span>        group.by <span class="op">=</span> <span class="st">"donor_id.and.stim"</span>, ncol <span class="op">=</span> <span class="fl">1</span><span class="op">)</span> </span></code></pre>
</div>
<figure><img src="../fig/section3-rendered-unnamed-chunk-13-1.png" style="display: block; margin: auto;" class="figure mx-auto d-block"></figure></div>
<div class="section level3">
<h3 id="step-6-creating-our-own-custom-visualisations-for-deg-analysis-between-cell-types-in-two-different-experimental-groups">Step 6: Creating our own custom visualisations for DEG analysis
between cell-types in two different experimental groups<a class="anchor" aria-label="anchor" href="#step-6-creating-our-own-custom-visualisations-for-deg-analysis-between-cell-types-in-two-different-experimental-groups"></a></h3>
<p>So far, we’ve been using functions wrapped within Seurat to plot and
visualise our data. But we’re not limited to those functions. Let’s use
the pheatmap (‘pretty heatmap’) package to visualise our DEGs with
adjusted p values &lt; 0.05.</p>
<p>First lets look at our significant DEGs as defined by our pseudobulk
approach</p>
<div class="codewrapper sourceCode" id="cb29">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">CD16.sig.markers</span> <span class="op">&lt;-</span> <span class="va">treatment.response.CD16.pseudo</span> <span class="op">%&gt;%</span> </span>
<span>  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu">filter</span><span class="op">(</span><span class="va">p_val_adj</span> <span class="op">&lt;</span> <span class="fl">0.05</span><span class="op">)</span> <span class="op">%&gt;%</span></span>
<span>  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu">mutate</span><span class="op">(</span>gene <span class="op">=</span> <span class="fu">rownames</span><span class="op">(</span><span class="va">.</span><span class="op">)</span><span class="op">)</span></span></code></pre>
</div>
<p>This is how we can pull our average (scaled) pseudobulk expression
values from our seurat obj:</p>
<div class="codewrapper sourceCode" id="cb30">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">ifnb.filtered</span><span class="op">$</span><span class="va">celltype.stim.donor_id</span> <span class="op">&lt;-</span> <span class="fu">paste0</span><span class="op">(</span><span class="va">ifnb.filtered</span><span class="op">$</span><span class="va">seurat_annotations</span>, <span class="st">"-"</span>,</span>
<span>                                               <span class="va">ifnb.filtered</span><span class="op">$</span><span class="va">stim</span>, <span class="st">"-"</span>, <span class="va">ifnb.filtered</span><span class="op">$</span><span class="va">donor_id</span><span class="op">)</span></span>
<span><span class="fu">Idents</span><span class="op">(</span><span class="va">ifnb.filtered</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="st">"celltype.stim.donor_id"</span></span>
<span></span>
<span><span class="va">all.sig.avg.Expression.mat</span> <span class="op">&lt;-</span> <span class="fu">AverageExpression</span><span class="op">(</span><span class="va">ifnb.filtered</span>, </span>
<span>                         features <span class="op">=</span> <span class="va">CD16.sig.markers</span><span class="op">$</span><span class="va">gene</span>, </span>
<span>                         layer <span class="op">=</span> <span class="st">'scale.data'</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>As of Seurat v5, we recommend using AggregateExpression to perform pseudo-bulk analysis.
This message is displayed once per session.</code></pre>
</div>
<div class="codewrapper sourceCode" id="cb32">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## As of Seurat v5, we recommend using AggregateExpression to perform pseudo-bulk analysis.</span></span>
<span><span class="co">## This message is displayed once per session.</span></span>
<span></span>
<span><span class="co"># View(all.sig.avg.Expression.mat %&gt;%</span></span>
<span><span class="co">#  as.data.frame())</span></span></code></pre>
</div>
<p>Now lets make sure we’re only using data from the CD16 cell type</p>
<div class="codewrapper sourceCode" id="cb33">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">CD16.sig.avg.Expression.mat</span> <span class="op">&lt;-</span> <span class="va">all.sig.avg.Expression.mat</span><span class="op">$</span><span class="va">RNA</span> <span class="op">%&gt;%</span></span>
<span>  <span class="fu">as.data.frame</span><span class="op">(</span><span class="op">)</span> <span class="op">%&gt;%</span></span>
<span>  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu">select</span><span class="op">(</span><span class="fu">starts_with</span><span class="op">(</span><span class="st">"CD16 Mono"</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># View(CD16.sig.avg.Expression.mat)</span></span></code></pre>
</div>
<p>Let’s finally view our heatmap with averaged pseudobulk scaled
expression values for our signficant DEGs</p>
<div class="codewrapper sourceCode" id="cb34">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">pheatmap</span><span class="fu">::</span><span class="fu">pheatmap</span><span class="op">(</span><span class="va">CD16.sig.avg.Expression.mat</span>,</span>
<span>         cluster_rows <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>         show_rownames <span class="op">=</span> <span class="cn">FALSE</span>, </span>
<span>         border_color <span class="op">=</span> <span class="cn">NA</span>, </span>
<span>         fontsize <span class="op">=</span> <span class="fl">10</span>, </span>
<span>         scale <span class="op">=</span> <span class="st">"row"</span>, </span>
<span>         fontsize_row <span class="op">=</span> <span class="fl">10</span>, </span>
<span>         height <span class="op">=</span> <span class="fl">20</span><span class="op">)</span></span></code></pre>
</div>
<figure><img src="../fig/section3-rendered-unnamed-chunk-17-1.png" style="display: block; margin: auto;" class="figure mx-auto d-block"></figure><p>The cool thing about pheatmap is that we can customise our heatmap
with added metadata</p>
<div class="codewrapper sourceCode" id="cb35">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">cluster_metadata</span> <span class="op">&lt;-</span> <span class="fu">data.frame</span><span class="op">(</span></span>
<span>  row.names <span class="op">=</span> <span class="fu">colnames</span><span class="op">(</span><span class="va">CD16.sig.avg.Expression.mat</span><span class="op">)</span></span>
<span><span class="op">)</span> <span class="op">%&gt;%</span> </span>
<span>  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu">mutate</span><span class="op">(</span></span>
<span>    Cell_Type <span class="op">=</span> <span class="st">"CD16 Mono"</span>,</span>
<span>    Treatment_Group <span class="op">=</span> <span class="fu">ifelse</span><span class="op">(</span><span class="fu">str_detect</span><span class="op">(</span><span class="fu">row.names</span><span class="op">(</span><span class="va">.</span><span class="op">)</span>, <span class="st">"STIM|CTRL"</span><span class="op">)</span>, </span>
<span>                      <span class="fu">str_extract</span><span class="op">(</span><span class="fu">row.names</span><span class="op">(</span><span class="va">.</span><span class="op">)</span>, <span class="st">"STIM|CTRL"</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="va">sig.DEG.heatmap</span> <span class="op">&lt;-</span> <span class="fu">pheatmap</span><span class="fu">::</span><span class="fu">pheatmap</span><span class="op">(</span><span class="va">CD16.sig.avg.Expression.mat</span>,</span>
<span>         cluster_rows <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>         show_rownames <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>         annotation <span class="op">=</span> <span class="va">cluster_metadata</span><span class="op">[</span>, <span class="fu">c</span><span class="op">(</span><span class="st">"Treatment_Group"</span>, <span class="st">"Cell_Type"</span><span class="op">)</span><span class="op">]</span>, </span>
<span>         border_color <span class="op">=</span> <span class="cn">NA</span>, </span>
<span>         fontsize <span class="op">=</span> <span class="fl">10</span>, </span>
<span>         scale <span class="op">=</span> <span class="st">"row"</span>, </span>
<span>         fontsize_row <span class="op">=</span> <span class="fl">10</span>, </span>
<span>         height <span class="op">=</span> <span class="fl">20</span>,</span>
<span>         annotation_names_col <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span></span>
<span><span class="va">sig.DEG.heatmap</span></span></code></pre>
</div>
<figure><img src="../fig/section3-rendered-unnamed-chunk-18-1.png" style="display: block; margin: auto;" class="figure mx-auto d-block"></figure><p>Lets save our heatmap using the grid library (very useful package!)
Code from: <a href="https://stackoverflow.com/questions/43051525/how-to-draw-pheatmap-plot-to-screen-and-also-save-to-file" class="external-link uri">https://stackoverflow.com/questions/43051525/how-to-draw-pheatmap-plot-to-screen-and-also-save-to-file</a></p>
<div class="codewrapper sourceCode" id="cb36">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">save_pheatmap_pdf</span> <span class="op">&lt;-</span> <span class="kw">function</span><span class="op">(</span><span class="va">x</span>, <span class="va">filename</span>, <span class="va">width</span><span class="op">=</span><span class="fl">7</span>, <span class="va">height</span><span class="op">=</span><span class="fl">7</span><span class="op">)</span> <span class="op">{</span></span>
<span>   <span class="fu">stopifnot</span><span class="op">(</span><span class="op">!</span><span class="fu">missing</span><span class="op">(</span><span class="va">x</span><span class="op">)</span><span class="op">)</span></span>
<span>   <span class="fu">stopifnot</span><span class="op">(</span><span class="op">!</span><span class="fu">missing</span><span class="op">(</span><span class="va">filename</span><span class="op">)</span><span class="op">)</span></span>
<span>   <span class="fu">pdf</span><span class="op">(</span><span class="va">filename</span>, width<span class="op">=</span><span class="va">width</span>, height<span class="op">=</span><span class="va">height</span><span class="op">)</span></span>
<span>   <span class="fu">grid</span><span class="fu">::</span><span class="fu">grid.newpage</span><span class="op">(</span><span class="op">)</span></span>
<span>   <span class="fu">grid</span><span class="fu">::</span><span class="fu">grid.draw</span><span class="op">(</span><span class="va">x</span><span class="op">$</span><span class="va">gtable</span><span class="op">)</span></span>
<span>   <span class="fu">dev.off</span><span class="op">(</span><span class="op">)</span></span>
<span><span class="op">}</span></span>
<span></span>
<span><span class="fu">save_pheatmap_pdf</span><span class="op">(</span><span class="va">sig.DEG.heatmap</span>, <span class="st">"sig_DEG_pseudo.pdf"</span><span class="op">)</span></span></code></pre>
</div>
<div id="challenge1" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<span class="callout-header">Challenge</span>
<div class="callout-inner">
<h3 class="callout-title">Challenge</h3>
<div class="callout-content">
<p>Do the same thing with significant DEGs from the sc approach. Do you
see any differences?</p>
</div>
</div>
</div>
<div id="accordionSolution1" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution1" aria-expanded="false" aria-controls="collapseSolution1">
  <h4 class="accordion-header" id="headingSolution1"> Show me the solution </h4>
</button>
<div id="collapseSolution1" class="accordion-collapse collapse" data-bs-parent="#accordionSolution1" aria-labelledby="headingSolution1">
<div class="accordion-body">
<p>Run the same code as above but with the
<code>treatment.response.CD16</code> variable instead of
<code>treatment.response.CD16.pseudo</code> variable.</p>
</div>
</div>
</div>
</div>
<p>You have now completed the workshop!</p>
<p>Hopefully you’ve found this workshop useful and you’re now feeling
more confident about using different integration and differential
expression approaches in single-cell data analysis.</p>
<div id="discussion5" class="callout discussion">
<div class="callout-square">
<i class="callout-icon" data-feather="message-circle"></i>
</div>
<span class="callout-header">Discussion</span>
<div class="callout-inner">
<div class="callout-content">
<p>What kind of analyses do you think we can do next after obtaining a
list of differentially expressed genes (DEGs) for each cell type and
treatment groups?</p>
</div>
</div>
</div>
<div id="keypoints1" class="callout keypoints">
<div class="callout-square">
<i class="callout-icon" data-feather="key"></i>
</div>
<span class="callout-header">Key Points</span>
<div class="callout-inner">
<div class="callout-content">
<ul><li>Collapse single-cell counts to pseudobulk per donor × condition ×
cell type.</li>
<li>Compare single-cell DE vs pseudobulk DE.</li>
</ul></div>
</div>
</div>
<!--
Place links that you need to refer to multiple times across pages here. Delete
any links that you are not going to use.
 -->
</div>
</section></div> <!-- / div.lesson-content -->
    </main><!-- / main#main-content.main-content --><nav class="bottom-pagination mx-md-4" aria-label="Previous and Next Chapter"><div class="d-block d-sm-block d-md-none">
        <a class="chapter-link" href="../instructor/section2.html"><i aria-hidden="true" class="small-arrow" data-feather="arrow-left"></i>Previous</a>
        <a class="chapter-link float-end" href="../instructor/index.html">Next<i aria-hidden="true" class="small-arrow" data-feather="arrow-right"></i></a>
      </div>
      <!-- content for large screens -->
      <div class="d-none d-sm-none d-md-block">
        <a class="chapter-link" href="../instructor/section2.html" rel="prev">
          <i aria-hidden="true" class="small-arrow" data-feather="arrow-left"></i>
          Previous: Section 2:
        </a>
        <a class="chapter-link float-end" href="../instructor/index.html" rel="next">
          Home
          <i aria-hidden="true" class="small-arrow" data-feather="arrow-right"></i>
        </a>
      </div>
    </nav></div> <!-- / div.primary-content.col-xs-12 -->
<!-- END:   inst/pkgdown/templates/content-instructor.html-->

      </div><!--/div.row-->
      		<footer class="row footer mx-md-3"><hr><div class="col-md-6">
        <p>This lesson is subject to the <a href="CODE_OF_CONDUCT.html">Code of Conduct</a></p>
        <p>

        <a href="https://example.com/template-source/lesson-example/edit/main/episodes/section3.Rmd" class="external-link">Edit on GitHub</a>

	
        | <a href="https://example.com/template-source/lesson-example/blob/main/CONTRIBUTING.md" class="external-link">Contributing</a>
        | <a href="https://example.com/template-source/lesson-example/" class="external-link">Source</a></p>
				<p><a href="https://example.com/template-source/lesson-example/blob/main/CITATION.cff" class="external-link">Cite</a> | <a href="mailto:team@carpentries.org">Contact</a> | <a href="https://carpentries.org/about/" class="external-link">About</a></p>
			</div>
			<div class="col-md-6">

        <p>Materials licensed under <a href="LICENSE.html">CC-BY 4.0</a> by the authors</p>

        <p>Template licensed under <a href="https://creativecommons.org/licenses/by-sa/4.0/" class="external-link">CC-BY 4.0</a> by <a href="https://carpentries.org/" class="external-link">The Carpentries</a></p>
        <p>Built with <a href="https://github.com/carpentries/sandpaper/tree/0.17.1" class="external-link">sandpaper (0.17.1)</a>, <a href="https://github.com/carpentries/pegboard/tree/0.7.9" class="external-link">pegboard (0.7.9)</a>, and <a href="https://github.com/carpentries/varnish/tree/1.0.7" class="external-link">varnish (1.0.7)</a></p>
			</div>
		</footer></div> <!-- / div.container -->
	<div id="to-top">
		<a href="#top">
      <i class="search-icon" data-feather="arrow-up" role="img" aria-label="Back To Top"></i><br><!-- <span class="d-none d-sm-none d-md-none d-lg-none d-xl-block">Back</span> To Top --><span class="d-none d-sm-none d-md-none d-lg-none d-xl-block">Back</span> To Top
		</a>
	</div>
  <script type="application/ld+json">
    {
  "@context": "https://schema.org",
  "@type": "LearningResource",
  "@id": "https://template-source.github.io/lesson-example/instructor/section3.html",
  "inLanguage": "en",
  "dct:conformsTo": "https://bioschemas.org/profiles/LearningResource/1.0-RELEASE",
  "description": "A Carpentries Lesson teaching foundational data and coding skills to researchers worldwide",
  "keywords": "software, data, lesson, The Carpentries",
  "name": "Section 3 : Differential Expression using a pseudobulk approach and DESeq2",
  "creativeWorkStatus": "active",
  "url": "https://template-source.github.io/lesson-example/instructor/section3.html",
  "identifier": "https://template-source.github.io/lesson-example/instructor/section3.html",
  "dateCreated": "2025-09-30",
  "dateModified": "2025-11-02",
  "datePublished": "2025-11-02"
}

  </script><script>
		feather.replace();
	</script></body></html><!-- END:   inst/pkgdown/templates/layout.html-->

